Fix logs (Other Error and forwarding related traces) (issue #1794)

Anjali-NEC opened the pull request:
Fix issue #1794

fgalan commented on doc/manuals/admin/logs.md:
```diff
@@ -224,6 +224,7 @@ Alarm conditions:
 | 3          | CRITICAL   | The following ERROR text appears in the 'msg' field: "Raising alarm DatabaseError: `<detail>`"                    | The following ERROR text appears in the 'msg' field: "Releasing alarm DatabaseError". Orion prints this trace when it detects that DB is ok again."                                                                                       | Database Error. The ...
 | 4          | WARNING    | The following WARN text appear in the 'msg' field: "Raising alarm BadInput `<ip>`: `<detail>`".                   | The following WARN text appears in the 'msg' field: "Releasing alarm BadInput `<ip>`", where <uip> is the same one that triggered the alarm. Orion prints this trace when it receives a correct request from that client.                 | Bad Input. The `<de ...
 | 5          | WARNING    | The following WARN text appears in the 'msg' field: "Raising alarm NotificationError `<url>`: `<detail>`".        | The following WARN text appears in the 'msg' field: "Releasing alarm NotificationError <url>", where <url> is the same one that triggered the alarm. Orion prints this trace when it successfully sent a notification to that URL.        | Notification Failur ...
+| 6          | WARNING    | The following WARN text appears in the 'msg' field: "Raising alarm ForwardingError `<url>`": `<detail>`".         | The following WARN text appears in the 'msg' field: "Releasing alarm ForwardingError <url>", where <url> is the same one that triggered the alarm. Orion prints this trace when it successfully interact with ContextProvider to that URL.| Forwarding Error. T ...
```
```suggestion
| 6          | WARNING    | The following WARN text appears in the 'msg' field: "Raising alarm ForwardingError `<url>`": `<detail>`".         | The following WARN text appears in the 'msg' field: "Releasing alarm ForwardingError `<url>`", where `<url>` is the same one that triggered the alarm. Orion prints this trace when it successfully interact with ContextProvider to that URL.| Forwarding Error. The `<detail>`text contains the detailed information.                                       | Orion is trying to interact with ContextProvider and some problem has occurred. It may be due to context provider response for forwarded query or update  is empty.
```

Anjali-NEC replied:
Updated

fgalan commented on doc/manuals/admin/logs.md:
```diff
@@ -224,6 +224,7 @@ Alarm conditions:
 | 3          | CRITICAL   | The following ERROR text appears in the 'msg' field: "Raising alarm DatabaseError: `<detail>`"                    | The following ERROR text appears in the 'msg' field: "Releasing alarm DatabaseError". Orion prints this trace when it detects that DB is ok again."                                                                                       | Database Error. The ...
```
An entry about the changes included in this PR, citing issue #1794, should be included in CHANGES_NEXT_RELEASE file.

Anjali-NEC replied:
Added entry about the changes in CHANGES_NEXT_RELEASE file.

fgalan commented:
Thanks for your contribution!

I have provided some comments in the PR. However, I understand is yet an ongoing work as although you have defined proper forwardingError() and forwardingErrorReset() methods, they are not actually used in any part of the code.

Hint: have a look to how "sibling methods" notificationError() and notificationErrorReset() are being used.

fgalan commented:
@Anjali-NEC I have seen recent activity in commit in this PR, but no comment on it...

Could you tell us a little bit about the status of the PR, please? Thanks!

Anjali-NEC commented:
Hi @fgalan Sir,

I have modified src/lib/ngsiNotify/senderThread.cpp and replaced the notificationError and notificationErrorReset with forwardingError and forwardingErrorReset as the notificationError  alarm was raised when notifications/forwarding receiver stop.

fgalan commented:
> I have modified src/lib/ngsiNotify/senderThread.cpp and replaced the notificationError and notificationErrorReset with forwardingError and forwardingErrorReset as the notificationError alarm was raised when notifications/forwarding receiver stop.

The idea is not to *replace* but to *create* now methods. From my original feedback at https://github.com/telefonicaid/fiware-orion/issues/1794#issuecomment-748178489

> Ensure that the alarm works properly in the program flow. I mean, **create** a couple of functions similar to `notificationError` and `notificationErrorReset` and use them in the right points. The xxxxError function has to be called in situations in which the forwarding has failed, the xxxxErrorReset function has to be called when a forwarding has gone well.

Maybe I'm missing something?

Anjali-NEC commented:
Hi @fgalan sir,

I have updated PR #3749. Please check PR and if it is ok then please merge the PR in master branch.

fgalan commented on doc/manuals/admin/logs.md:
```diff
@@ -222,8 +222,10 @@ Alarm conditions:
 | 3          | CRITICAL   | The following ERROR text appears in the 'msg' field: "Raising alarm DatabaseError: `<detail>`"                    | The following ERROR text appears in the 'msg' field: "Releasing alarm DatabaseError". Orion prints this trace when it detects that DB is ok again."                                                                                       | Database Error. The ...
-| 4          | WARNING    | The following WARN text appear in the 'msg' field: "Raising alarm BadInput `<ip>`: `<detail>`".                   | The following WARN text appears in the 'msg' field: "Releasing alarm BadInput `<ip>`", where `<ip>` is the same one that triggered the alarm. Orion prints this trace when it receives a correct request from that client.                 | Bad Input. The `<d ...
-| 5          | WARNING    | The following WARN text appears in the 'msg' field: "Raising alarm NotificationError `<url>`: `<detail>`".        | The following WARN text appears in the 'msg' field: "Releasing alarm NotificationError `<url>`", where `<url>` is the same one that triggered the alarm. Orion prints this trace when it successfully sent a notification to that URL.        | Notification Fa ...
+| 4          | WARNING    | The following WARN text appear in the 'msg' field: "Raising alarm BadInput `<ip>`: `<detail>`".                   | The following WARN text appears in the 'msg' field: "Releasing alarm BadInput `<ip>`", where <uip> is the same one that triggered the alarm. Orion prints this trace when it receives a correct request from that client.                 | Bad Input. The `<de ...
```
This line (alarm type 4) shouldn't be changed by your PR.

Anjali-NEC replied:
Fixed in ca4d0d9

fgalan replied:
I'm afraid this is not correct, as if you look to "Files" for the full diff in the PR you would see red and green blocks with regards to alarm type 4. That probes that line has been "touched" in the PR. The line should appear without any red or green background colour.

fgalan commented on doc/manuals/admin/logs.md:
```diff
@@ -222,8 +222,10 @@ Alarm conditions:
-| 4          | WARNING    | The following WARN text appear in the 'msg' field: "Raising alarm BadInput `<ip>`: `<detail>`".                   | The following WARN text appears in the 'msg' field: "Releasing alarm BadInput `<ip>`", where `<ip>` is the same one that triggered the alarm. Orion prints this trace when it receives a correct request from that client.                 | Bad Input. The `<d ...
-| 5          | WARNING    | The following WARN text appears in the 'msg' field: "Raising alarm NotificationError `<url>`: `<detail>`".        | The following WARN text appears in the 'msg' field: "Releasing alarm NotificationError `<url>`", where `<url>` is the same one that triggered the alarm. Orion prints this trace when it successfully sent a notification to that URL.        | Notification Fa ...
+| 4          | WARNING    | The following WARN text appear in the 'msg' field: "Raising alarm BadInput `<ip>`: `<detail>`".                   | The following WARN text appears in the 'msg' field: "Releasing alarm BadInput `<ip>`", where <uip> is the same one that triggered the alarm. Orion prints this trace when it receives a correct request from that client.                 | Bad Input. The `<de ...
+| 5          | WARNING    | The following WARN text appears in the 'msg' field: "Raising alarm NotificationError `<url>`: `<detail>`".        | The following WARN text appears in the 'msg' field: "Releasing alarm NotificationError <url>", where <url> is the same one that triggered the alarm. Orion prints this trace when it successfully sent a notification to that URL.        | Notification Failur ...
```
This line (alarm type 5) shouldn't be changed by your PR.

Anjali-NEC replied:
Fixed in ca4d0d9

fgalan replied:
I'm afraid this is not correct, as if you look to "Files" for the full diff in the PR you would see red and green blocks with regards to alarm type 4. That probes that line has been "touched" in the PR. The line should appear without any red or green background colour.

Anjali-NEC replied:
>I'm afraid this is not correct, as if you look to "Files" for the full diff in the PR you would see red and green blocks with regards to alarm type 4. That probes that line has been "touched" in the PR. The line should appear without any red or green background colour.

Fixed in 32c8f91

fgalan commented on doc/manuals/admin/logs.md:
```diff
@@ -222,8 +222,10 @@ Alarm conditions:
-| 5          | WARNING    | The following WARN text appears in the 'msg' field: "Raising alarm NotificationError `<url>`: `<detail>`".        | The following WARN text appears in the 'msg' field: "Releasing alarm NotificationError `<url>`", where `<url>` is the same one that triggered the alarm. Orion prints this trace when it successfully sent a notification to that URL.        | Notification Fa ...
+| 4          | WARNING    | The following WARN text appear in the 'msg' field: "Raising alarm BadInput `<ip>`: `<detail>`".                   | The following WARN text appears in the 'msg' field: "Releasing alarm BadInput `<ip>`", where <uip> is the same one that triggered the alarm. Orion prints this trace when it receives a correct request from that client.                 | Bad Input. The `<de ...
+| 5          | WARNING    | The following WARN text appears in the 'msg' field: "Raising alarm NotificationError `<url>`: `<detail>`".        | The following WARN text appears in the 'msg' field: "Releasing alarm NotificationError <url>", where <url> is the same one that triggered the alarm. Orion prints this trace when it successfully sent a notification to that URL.        | Notification Failur ...
+| 6          | WARNING    | The following WARN text appears in the 'msg' field: "Raising alarm ForwardingError `<url>`": `<detail>`".         | The following WARN text appears in the 'msg' field: "Releasing alarm ForwardingError `<url>`", where `<url>` is the same one that triggered the alarm. Orion prints this trace when it successfully interact with ContextProvider to that URL.| Forwarding Erro ...
```
New entry for alarm type 6 looks fine. Just a little addition:

```suggestion
| 6          | WARNING    | The following WARN text appears in the 'msg' field: "Raising alarm ForwardingError `<url>`": `<detail>`".         | The following WARN text appears in the 'msg' field: "Releasing alarm ForwardingError `<url>`", where `<url>` is the same one that triggered the alarm. Orion prints this trace when it successfully interact with ContextProvider to that URL.| Forwarding Error. The `<detail>`text contains the detailed information.                                       | Orion is trying to interact with ContextProvider and some problem has occurred. It may be due to context provider response for forwarded query or update is empty. No specific action has to be performed at Orion Context Broker service.
```

Anjali-NEC replied:
Fixed in ca4d0d9

fgalan commented on src/lib/ngsiNotify/senderThread.cpp:
```diff
@@ -99,6 +100,7 @@ void* startSenderThread(void* p)
       else
       {
         alarmMgr.notificationError(url, "notification failure for sender-thread: " + out);
+        alarmMgr.forwardingError(url, "forwarding failure for sender-thread: " + out);
```
This is not correct...

The sender thread is dealing with either a notification or a forwarding, but not both at the same time. So if should be something like this:

```
if (<some condition that when true means the sender thread is dealing with a notification>)
{
 alarmMgr.notificationError(url, "notification failure for sender-thread: " + out);
}
else
{
  alarmMgr.forwardingError(url, "forwarding failure for sender-thread: " + out);
}
```

The key point here is the condition to distinguish between one case or another... I don't which one should be used. Any idea?

fgalan replied:
(The same applies to the alarm reset logic, around line 93)

fgalan commented:
> Hi @fgalan sir,
> 
> I have updated PR #3749. Please check PR and if it is ok then please merge the PR in master branch.

I have done a new review and provide new feedback. In general, I think the PR looks good but a key aspect, commented [here](https://github.com/telefonicaid/fiware-orion/pull/3749#discussion_r611435384) needs to be solved.

In addition, it would be great if you could upgrade this PR branch with master and solve the existing conflict in CHANGES_NEXT_RELEASE, please.

Anjali-NEC commented:
> > Hi @fgalan sir,
> > I have updated PR #3749. Please check PR and if it is ok then please merge the PR in master branch.
> 
> I have done a new review and provide new feedback. In general, I think the PR looks good but a key aspect, commented [here](https://github.com/telefonicaid/fiware-orion/pull/3749#discussion_r611435384) needs to be solved.
> 
> In addition, it would be great if you could upgrade this PR branch with master and solve the existing conflict in CHANGES_NEXT_RELEASE, please.

Updated PR #3749 as per comment

fgalan commented:
> 
> 
> > > Hi @fgalan sir,
> > > I have updated PR #3749. Please check PR and if it is ok then please merge the PR in master branch.
> > 
> > 
> > I have done a new review and provide new feedback. In general, I think the PR looks good but a key aspect, commented [here](https://github.com/telefonicaid/fiware-orion/pull/3749#discussion_r611435384) needs to be solved.
> > In addition, it would be great if you could upgrade this PR branch with master and solve the existing conflict in CHANGES_NEXT_RELEASE, please.
> 
> Updated PR #3749 as per comment

I'm still seeing yet a couple of opened comment threads. In particular:

* https://github.com/telefonicaid/fiware-orion/pull/3749#pullrequestreview-633267353
* https://github.com/telefonicaid/fiware-orion/pull/3749#pullrequestreview-633267464
* https://github.com/telefonicaid/fiware-orion/pull/3749#discussion_r611427211
* https://github.com/telefonicaid/fiware-orion/pull/3749#discussion_r611435384

Note that for closing a thread it should end in a comment with the commit reference in which you have addressed the comment or the NTC acronym, explaining why the suggested modification cannot be applied. More detail on the PR procedure [here](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/contribution_guidelines.md#pull-request-protocol)).

The fourth one is specially important. Could you elaborate on the condition (i.e. `<some condition that when true means the sender thread is dealing with a notification>`) you have used to distinguish between both cases, please?

Anjali-NEC commented:
> The fourth one is specially important. Could you elaborate on the condition (i.e. `<some condition that when true means the sender thread is dealing with a notification>`) you have used to distinguish between both cases, please?

I have used the condition     `if ((cleanPayload == NULL) || (cleanPayload[0] == 0))` refer from the code used in postQueryContext/postUpdateContext https://github.com/telefonicaid/fiware-orion/blob/master/src/lib/serviceRoutines/postUpdateContext.cpp line number: 261
to check  context provider response for forwarded query or update is empty if it is empty then it will raised forwardingError alarm
else it will raised notificationError alarm.

fgalan commented:
I have had a closer look to the pending issue in this PR.

It seems that the logic in src/lib/ngsiNotify/senderThread.cpp is only used by notification logic. Note that the forwarding logic is invoking directly to httpRequestSend() in updateForward() function (postUpdateContext.cpp) and queryForward() function (postQueryCotnext.cpp).

Thus, I think the actions should be done in this PR are the following:

1. Revert src/lib/ngsiNotify/senderThread.cpp to its version in master branch, as no change is needed there.
2. Include the alarm/raise logic for forwarding alarms in updateForward()
3. Include the alarm/raise logic for forwarding alarms in queryForward()

For items 2 and 3 it should be easy.

```
  r = httpRequestSend(fromIp,   // thread variable
        ...
                      mimeType);

  if (r != 0)
  {
    ...
    //  <code for raising alarm here>
    return false;
  }
  // <code for releasing alarm here>
```

I hope this new feedback helps.

fgalan commented on CHANGES_NEXT_RELEASE:
```diff
@@ -1,4 +1,6 @@
+- Fix logs (Other Error and forwarding related traces) (issue #1794)
```
```suggestion
- Add: new alarm (id number 6) for forwarding errors (#1794)
- Fix: use "Forwarding Error" instead of "Other Error" in WARN traces (#1794)
```

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3749/commits/49298094b6110acc3cb09e5c5a23442fa93ab5f7

Anjali-NEC commented:
I have tried to Include the alarm/raise logic for forwarding alarms in updateForward() and queryForward() as suggested by you in comment https://github.com/telefonicaid/fiware-orion/pull/3749#issuecomment-843988019. and try to  verify the forwarding alarm in logs but I could not found any logs of forwarding alarm. I am following the steps to verify the forwarding alarm logs as suggest by you in comment https://github.com/telefonicaid/fiware-orion/issues/1794#issuecomment-756194674 

Please guide me further.

fgalan commented:
> I have tried to Include the alarm/raise logic for forwarding alarms in updateForward() and queryForward() as suggested by you in comment [#3749 (comment)](https://github.com/telefonicaid/fiware-orion/pull/3749#issuecomment-843988019). and try to verify the forwarding alarm in logs but I could not found any logs of forwarding alarm. I am following the steps to verify the forwarding alarm logs as suggest by you in comment [#1794 (comment)](https://github.com/telefonicaid/fiware-orion/issues/1794#issuecomment-756194674)

To do this test the best option is to use two CBs processes: one of them acting as CB itself (let's say running in port 1026) and other CB acting as Context Provider (let's say running in port 1027). Use `-logLevel INFO` to get useful log traces.

Then, the test to do is as follows:

1. Create an entity E in the CPr (create entity request on port 1027)
2. Register CPr on CB (create registration on port 1026)
3. Query E at CB. The query would be forwarded to the CPr. Everything ok.
4. Turn down CPr process
5. Repeat step 3 query. The query is forwarded but now, given the CPrs is down, a connection error ocurrs and you should see a forwarding alarm **raise** message in the logs.
6. Turn up CPr process
7. Repeat step 3 query. Now the query forwarding would work again and you should see a forwarding alarm **release** message in the logs.

And a similar test for update forwarding.

Existing tests like the ones test/functionalTest/cases/3068_cprs_full_functional_v2/ could help to understand this kind of testing setup.

Anyway, if testing this is difficult for you, I can try to help. Update your PR with the modifications you has done recently and I can provide more feedback.

Anjali-NEC commented:
Hi @fgalan Sir,

Thank you for your guidance and help. I have updated my PR in my latest commit https://github.com/telefonicaid/fiware-orion/pull/3749/commits/cf65e3f9c9f785761647617bca0fc91c57f2bb9b

fgalan commented on src/lib/serviceRoutines/postQueryContext.cpp:
```diff
@@ -255,6 +255,7 @@ static bool queryForward
   //
   std::string     out;
   int             r;
+  std::string     url;
```
It seems this variable is not initialized...

Looking to similar code in senderThread.cpp I think this could be used:

```
url = ip + ":" + port + resource;
```

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3749/commits/d4e75f554d1327a17ec1235529378567fd5c4454

fgalan commented on src/lib/serviceRoutines/postUpdateContext.cpp:
```diff
@@ -190,6 +190,7 @@ static bool updateForward
 
   ciP->outMimeType  = outMimeType;
   cleanPayload      = (char*) payload.c_str();
+  std::string         url;
```
Same comment here: https://github.com/telefonicaid/fiware-orion/pull/3749/files#r636880073

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3749/commits/d4e75f554d1327a17ec1235529378567fd5c4454

fgalan commented on src/lib/serviceRoutines/postQueryContext.cpp:
```diff
@@ -255,6 +255,8 @@ static bool queryForward
   std::string     out;
   int             r;
+  std::string     url;
+  char            portV[STRING_SIZE_FOR_INT];
```
Where this portV variable is set? Maybe you need something like:

```
snprintf(portV, sizeof(portV), "%d", port);
```

Please have a look to similar code at senderThread.cpp and check how it is done there.

Anjali-NEC replied:
Added portV variable in line number 270 in src/lib/serviceRoutines/postQueryContext.cpp and similarly in line number 208 src/lib/serviceRoutines/postUpdateContext.cpp

Please check my latest commit https://github.com/telefonicaid/fiware-orion/pull/3749/commits/d4e75f554d1327a17ec1235529378567fd5c4454 and let me know if I am missing something.

fgalan replied:
You are right. I missed that `sprintf` in you last commit, sorry...

NTC

fgalan commented:
Test done:

Running CB the following way:

```
contextBroker -fg -logLevel INFO -logForHumans -harakiri
```

Running CP the following way:

```
contextBroker -fg -logLevel INFO -logForHumans -harakiri -port 1027 -db orion-cpr
```

---------------------

Test for **query forwarding**

1. Create an entity E in the CPr using:

```
curl localhost:1027/v2/entities -s -S -H 'Content-Type: application/json' -d @- <<EOF
{
  "id": "E",
  "type": "T",
  "temperature": {
    "value": 23,
    "type": "Float"
  }
}
EOF
```
2. Register CPr on CB:

```
curl localhost:1026/v2/registrations -s -S -H 'Content-Type: application/json' -H 'Accept: application/json' -d @-  <<EOF
{
  "dataProvided": {
    "entities": [
      {
        "id": "E",
        "type": "T"
      }
    ],
    "attrs": [
      "temperature"
    ]
  },
  "provider": {
    "http": {
      "url": "http://localhost:1027/v2"
    }
  }
}
EOF
```

3. Query E at CB.

```
curl -s -S localhost:1026/v2/entities/E | python -mjson.tool
```

we see at the logs CB logs:

```
INFO@2021-05-24T11:18:47.676Z  logTracing.cpp[146]: Starting forwarding for GET /v2/entities/E
INFO@2021-05-24T11:18:47.683Z  logTracing.cpp[212]: Request forwarded (regId: 60ab8b5853d65101092c46eb): POST http://localhost:1027/v2/op/query, request payload (60 bytes): {"entities":[{"id":"E","type":"T"}],"attrs":["temperature"]}, response payload (79 bytes): [{"id":"E","type":"T","temperature":{"type":"Float","value":23,"metadata":{}}}], response code: 200
INFO@2021-05-24T11:18:47.683Z  logTracing.cpp[79]: Request received: GET /v2/entities/E, response code: 200
```

we see at the CPr logs:

```
NFO@2021-05-24T11:18:47.684Z  logTracing.cpp[130]: Request received: POST /v2/op/query, request payload (60 bytes): {"entities":[{"id":"E","type":"T"}],"attrs":["temperature"]}, response code: 200
```

so this is **OK**.

4. Now I turn down CPrs process and repeat previous step. We see at CB logs:

```
INFO@2021-05-24T11:20:32.175Z  logTracing.cpp[146]: Starting forwarding for GET /v2/entities/E
ERROR@2021-05-24T11:20:32.180Z  postQueryContext.cpp[295]: Runtime Error (error 'Couldn't connect to server' forwarding 'Query' to providing application)
INFO@2021-05-24T11:20:32.180Z  logTracing.cpp[212]: Request forwarded (regId: 60ab8b5853d65101092c46eb): POST http://localhost:1027/v2/op/query, request payload (60 bytes): {"entities":[{"id":"E","type":"T"}],"attrs":["temperature"]}, response payload (0 bytes): , response code: Couldn't connect to server
WARN@2021-05-24T11:20:32.181Z  AlarmManager.cpp[481]: Raising alarm forwardingError localhost:1027/v2/op/query: forwarding failure for sender-thread: Couldn't connect to server
INFO@2021-05-24T11:20:32.181Z  logTracing.cpp[79]: Request received: GET /v2/entities/E, response code: 404
```

We see how alarm is raised with `Raising alarm forwardingError localhost:1027/v2/op/query: forwarding failure for sender-thread: Couldn't connect to server`. Thus, this is also **OK**.

5. Now I start again the CPr process and repeat previous step. We see at CB logs:

```
INFO@2021-05-24T11:21:58.360Z  logTracing.cpp[146]: Starting forwarding for GET /v2/entities/E
WARN@2021-05-24T11:21:58.371Z  AlarmManager.cpp[508]: Releasing alarm forwardingError localhost:1027/v2/op/query
INFO@2021-05-24T11:21:58.371Z  logTracing.cpp[212]: Request forwarded (regId: 60ab8b5853d65101092c46eb): POST http://localhost:1027/v2/op/query, request payload (60 bytes): {"entities":[{"id":"E","type":"T"}],"attrs":["temperature"]}, response payload (79 bytes): [{"id":"E","type":"T","temperature":{"type":"Float","value":23,"metadata":{}}}], response code: 200
INFO@2021-05-24T11:21:58.371Z  logTracing.cpp[79]: Request received: GET /v2/entities/E, response code: 200
```

We see how alarm is released with `Releasing alarm forwardingError localhost:1027/v2/op/query`. Thus, this is also **OK**.

---------------

Now let's continue (same entity and registration) with the **update forwarding case**.

6. Update E at CB.

```
curl -X PATCH localhost:1026/v2/entities/E/attrs?type=T -s -S -H 'Content-Type: application/json' -d @- <<EOF
{
  "temperature": {
    "value": 42,
    "type": "Float"
  }
}
EOF
```

we see at CB logs:

```
INFO@2021-05-24T11:24:57.979Z  logTracing.cpp[146]: Starting forwarding for PATCH /v2/entities/E/attrs?type=T
INFO@2021-05-24T11:24:57.987Z  logTracing.cpp[212]: Request forwarded (regId: 60ab8b5853d65101092c46eb): POST http://localhost:1027/v2/op/update, request payload (114 bytes): {"entities":[{"id":"E","type":"T","temperature":{"type":"Float","value":42,"metadata":{}}}],"actionType":"update"}, response payload (0 bytes): , response code: 204
INFO@2021-05-24T11:24:57.988Z  logTracing.cpp[130]: Request received: PATCH /v2/entities/E/attrs?type=T, request payload (58 bytes): {  "temperature": {    "value": 42,    "type": "Float"  }}, response code: 204
```

we see at CPr logs:

```
INFO@2021-05-24T11:24:57.989Z  logTracing.cpp[130]: Request received: POST /v2/op/update, request payload (114 bytes): {"entities":[{"id":"E","type":"T","temperature":{"type":"Float","value":42,"metadata":{}}}],"actionType":"update"}, response code: 204
```

so this is **OK*.

7. Now I turn down CPrs process and repeat previous step. We see at CB logs:

```
INFO@2021-05-24T11:26:06.942Z  logTracing.cpp[146]: Starting forwarding for PATCH /v2/entities/E/attrs?type=T
ERROR@2021-05-24T11:26:06.947Z  postUpdateContext.cpp[234]: Runtime Error (error 'Couldn't connect to server' forwarding 'Update' to providing application)
INFO@2021-05-24T11:26:06.947Z  logTracing.cpp[212]: Request forwarded (regId: 60ab8b5853d65101092c46eb): POST http://localhost:1027/v2/op/update, request payload (114 bytes): {"entities":[{"id":"E","type":"T","temperature":{"type":"Float","value":42,"metadata":{}}}],"actionType":"update"}, response payload (0 bytes): , response code: Couldn't connect to server
WARN@2021-05-24T11:26:06.947Z  AlarmManager.cpp[481]: Raising alarm forwardingError localhost:1027/v2/op/update: forwarding failure for sender-thread: Couldn't connect to server
INFO@2021-05-24T11:26:06.948Z  logTracing.cpp[130]: Request received: PATCH /v2/entities/E/attrs?type=T, request payload (58 bytes): {  "temperature": {    "value": 42,    "type": "Float"  }}, response code: 404
```

We see how alarm is raised with `Raising alarm forwardingError localhost:1027/v2/op/update: forwarding failure for sender-thread: Couldn't connect to server`. Thus, this is also **OK**.

8. Now I start again the CPr process and repeat previous step. We see at CB logs:

```
INFO@2021-05-24T11:28:53.376Z  logTracing.cpp[146]: Starting forwarding for PATCH /v2/entities/E/attrs?type=T
WARN@2021-05-24T11:28:53.384Z  AlarmManager.cpp[508]: Releasing alarm forwardingError localhost:1027/v2/op/update
INFO@2021-05-24T11:28:53.384Z  logTracing.cpp[212]: Request forwarded (regId: 60ab8b5853d65101092c46eb): POST http://localhost:1027/v2/op/update, request payload (114 bytes): {"entities":[{"id":"E","type":"T","temperature":{"type":"Float","value":42,"metadata":{}}}],"actionType":"update"}, response payload (0 bytes): , response code: 204
INFO@2021-05-24T11:28:53.385Z  logTracing.cpp[130]: Request received: PATCH /v2/entities/E/attrs?type=T, request payload (58 bytes): {  "temperature": {    "value": 42,    "type": "Float"  }}, response code: 204
```

We see how alarm is released with `Releasing alarm forwardingError localhost:1027/v2/op/update`. Thus, this is also **OK**.

--------------------

In sum, all the cases are **OK**

fgalan commented:
I have found an small error doing my tests.

First time alarm is raised is fine. I mean:

```
INFO@2021-05-24T11:36:23.556Z  logTracing.cpp[146]: Starting forwarding for PATCH /v2/entities/E/attrs?type=T
ERROR@2021-05-24T11:36:23.561Z  postUpdateContext.cpp[234]: Runtime Error (error 'Couldn't connect to server' forwarding 'Update' to providing application)
INFO@2021-05-24T11:36:23.561Z  logTracing.cpp[212]: Request forwarded (regId: 60ab8b5853d65101092c46eb): POST http://localhost:1027/v2/op/update, request payload (114 bytes): {"entities":[{"id":"E","type":"T","temperature":{"type":"Float","value":42,"metadata":{}}}],"actionType":"update"}, response payload (0 bytes): , response code: Couldn't connect to server
WARN@2021-05-24T11:36:23.561Z  AlarmManager.cpp[481]: Raising alarm forwardingError localhost:1027/v2/op/update: forwarding failure for sender-thread: Couldn't connect to server
INFO@2021-05-24T11:36:23.561Z  logTracing.cpp[130]: Request received: PATCH /v2/entities/E/attrs?type=T, request payload (58 bytes): {  "temperature": {    "value": 42,    "type": "Float"  }}, response code: 404
```

However, if a given alarm is raised and the request causing the alarm is sent again we see:

```
INFO@2021-05-24T11:36:50.613Z  logTracing.cpp[146]: Starting forwarding for PATCH /v2/entities/E/attrs?type=T
WARN@2021-05-24T11:36:50.626Z  AlarmManager.cpp[508]: Releasing alarm forwardingError localhost:1027/v2/op/update
ERROR@2021-05-24T11:36:50.627Z  postUpdateContext.cpp[234]: Runtime Error (error 'Couldn't connect to server' forwarding 'Update' to providing application)
INFO@2021-05-24T11:36:50.627Z  logTracing.cpp[212]: Request forwarded (regId: 60ab8b5853d65101092c46eb): POST http://localhost:1027/v2/op/update, request payload (114 bytes): {"entities":[{"id":"E","type":"T","temperature":{"type":"Float","value":42,"metadata":{}}}],"actionType":"update"}, response payload (0 bytes): , response code: Couldn't connect to server
WARN@2021-05-24T11:36:50.627Z  AlarmManager.cpp[481]: Raising alarm forwardingError localhost:1027/v2/op/update: forwarding failure for sender-thread: Couldn't connect to server
INFO@2021-05-24T11:36:50.627Z  logTracing.cpp[130]: Request received: PATCH /v2/entities/E/attrs?type=T, request payload (58 bytes): {  "temperature": {    "value": 42,    "type": "Float"  }}, response code: 404
```

That's not correct: we shoudn't see that release+raise effect. In this case (assuming `-relogAlarms` is not used) no new release/raise trace should be print.

However, I think is easy to solve. Just move the release the forwardingErrorReset() call to an `else` block in the `if` clause that checks the result of the forwarding. I mean:

```
  if (r != 0)
  {
   ... 
  }
  else
  {
    alarmMgr.forwardingErrorReset(url);
  }
```

Anjali-NEC commented:
> However, I think is easy to solve. Just move the release the forwardingErrorReset() call to an `else` block in the `if` clause that checks the result of the forwarding. I mean:
> 
> ```
>   if (r != 0)
>   {
>    ... 
>   }
>   else
>   {
>     alarmMgr.forwardingErrorReset(url);
>   }
> ```

Fixed in https://github.com/telefonicaid/fiware-orion/pull/3749/commits/0c85cf127efad84c627a4e1a17c81be2c35d1118

fgalan commented on src/lib/serviceRoutines/postQueryContext.cpp:
```diff
@@ -282,13 +287,18 @@ static bool queryForward
                       noHeaders,
                       mimeType);
-
+                
```
```suggestion

```

(Avoid whitespace in blank line))

Anjali-NEC replied:
Fixed in https://github.com/telefonicaid/fiware-orion/pull/3749/commits/4c99163abe5ff68a1ee5eb612d0e9a8c74de5958

fgalan reviewed (approved):
LGTM

@Anjali-NEC thank you for your contribution!
